lib/excon/socket.rb

```diff
@@ -237,18 +269,33 @@ def read_nonblock(max_length)
         @backend_eof = true
       end
 
-      ret = if max_length
+      if max_length
         if @read_buffer.empty?
-          nil # EOF met at beginning
+          # EOF met at beginning
+          @eof = @backend_eof
+          nil
         else
-          @read_buffer.slice!(0, max_length)
+          start = @read_offset
+          readable_bytes = buffer_length
+
+          # Ensure that we can seek backwards when reading until a terminator string.
+          # The read offset must never point past the end of the read buffer.
+          @read_offset += max_length > readable_bytes ? readable_bytes : max_length
```

Comment:
Relating to the comment below, if `buffer_length` was referred to as something like `remaining_buffer`, I think we could probably just refer to it directly here and things would be relatively clear (as opposed to assigning it to `readable_bytes` in order to provide clarity, or at least I'm not immediately seeing another reason for that particular allocation).

Reply:
That makes sense to me.

Reply:
However, I don't think a single method call is going to be the differentiator here. I changed the code to use the method instead of a variable on the stack.

Reply:
yeah, probably would really be fine either way.